File: Leetcode/Algorithm/python/01909-Remove One Element to Make the Array Strictly Increasing.py
INF = 10 ** 10

class Solution(object):
    def canBeIncreasing(self, nums):
        count = 0
        nums = [-INF] + nums
        for i, num in enumerate(nums):
            if i == 0:
                continue
            if num <= nums[i - 1]:
                if count > 0:
                    return False
                if nums[i] <= nums[i - 2]:
                    nums[i] = nums[i - 1]
                    count += 1
                else:
                    count += 1
            # print i, count
            if count > 1:
                return False
        return True

File: Leetcode/Algorithm/cpp/01909-Remove One Element to Make the Array Strictly Increasing.cc
class Solution {
public:
    bool canBeIncreasing(vector<int>& nums) {
        vector<int> st;
        for (auto num: nums) {
            if (st.empty() || *st.rbegin() < num) {
                st.push_back(num);
                continue;
            }
            auto iter = lower_bound(st.begin(), st.end(), num);
            *iter = num;
        }
        return st.size() >= nums.size() - 1;
    }
};
